### PR TITLE
Layer composition rendering using RenderPass, optimization

### PR DESCRIPTION
**Pull request by mvaligursky**
related to issue: https://github.com/playcanvas/engine/issues/4271
follow up on PR: https://github.com/playcanvas/engine/pull/4272
partially implements https://github.com/playcanvas/engine/issues/2396

new (private) APIs:
```
GraphicsDevice.startPass(renderPass);
GraphicsDevice.endPass(renderPass);

```
This PR further implements the FrameGraph / RenderPass rendering architecture, specifically:
- the frame graph tracks load and store operations for render targets, in order to avoid some wasted loading and storing those to tiled memory
- the frame graph optimizes resolve and generate mipmaps operations as well

The benefits are mostly visible in projects that render to textures, especially if multi-sampling is used, including rendering to cubemaps.

GPU Performance improvement on some engine examples:
Screenshot 2022-06-10 at 09 55 41

Other changes
- log tracing added for render passes:
Screenshot 2022-06-10 at 09 51 28

- log tracing added for texture / render target allocation / deallocation
Screenshot 2022-06-10 at 09 54 17

### CODE REVIEW

File: src/graphics/graphics-device.js [hunk 1]
```diff
@@ -75,6 +75,12 @@ class GraphicsDevice extends EventHandler {
      */
     precision;
 
+    /** @type {RenderTarget} */
```
Comment (slimbuck) [inline src/graphics/graphics-device.js:78]: perhaps an explanation here on how this member is used by render-pass system?

File: src/graphics/render-pass.js [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+     * called only for render passes which use render target, or passes which render directly into the default
+     * framebuffer, in which case a null or undefined render target is expected.
+     */
+    init(renderTarget) {
+
+        // null represents the default framebuffer
+        this.renderTarget = renderTarget || null;
+
+        // allocate ops only when render target is used
+        if (this.renderTarget !== undefined) {
```
Comment (slimbuck) [inline src/graphics/render-pass.js:154]:
```suggestion
        if (this.renderTarget !== null) {
```

File: src/graphics/render-pass.js [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            this.colorOps = new ColorAttachmentOps();
+            this.depthStencilOps = new DepthStencilAttachmentOps();
+
+            // defaults depend on multisampling
+            this.samples = Math.max(this.renderTarget ? this.renderTarget.samples : this.device.samples, 1);
+            if (this.samples === 1) {
+                this.colorOps.store = true;
```
Comment (slimbuck) [inline src/graphics/render-pass.js:161]: could you explain why this logic?

File: src/graphics/render-pass.js [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        this.depthStencilOps.clearStencil = true;
+    }
+
+    /**
+     * Render the render pass
+     */
+    render() {
+
+        const device = this.device;
+        const realPass = this.renderTarget !== undefined;
```
Comment (slimbuck) [inline src/graphics/render-pass.js:208]:
```suggestion
        const realPass = this.renderTarget !== null;
```

File: src/graphics/webgl/webgl-graphics-device.js [hunk 5]
```diff
@@ -1235,14 +1238,133 @@ class WebglGraphicsDevice extends GraphicsDevice {
...
+            }
+
+            // resolve the color buffer
+            if (renderPass.colorOps.resolve) {
+                if (this.webgl2 && renderPass.samples > 1 && target.autoResolve) {
+                    target.resolve(true, false);
+                }
+            }
+
+            // generate mimaps
```
Comment (slimbuck) [inline src/graphics/webgl/webgl-graphics-device.js:1342]:
```suggestion
            // generate mipmaps
```

File: src/graphics/webgl/webgl-graphics-device.js [hunk 6]
```diff
@@ -1235,14 +1238,133 @@ class WebglGraphicsDevice extends GraphicsDevice {
...
+                // TODO: we could handle viewport invalidation as well
+                if (renderPass.fullSizeClearRect) {
+                    gl.invalidateFramebuffer(gl.DRAW_FRAMEBUFFER, invalidateAttachments);
+                }
+            }
+
+            // resolve the color buffer
+            if (renderPass.colorOps.resolve) {
+                if (this.webgl2 && renderPass.samples > 1 && target.autoResolve) {
+                    target.resolve(true, false);
```
Comment (slimbuck) [inline src/graphics/webgl/webgl-graphics-device.js:1338]: false for depth because webgpu?

File: src/scene/frame-graph.js [hunk 7]
```diff
@@ -14,43 +14,154 @@ class FrameGraph {
...
+    compile() {
+
+        const renderTargetMap = this.renderTargetMap;
         const renderPasses = this.renderPasses;
         for (let i = 0; i < renderPasses.length; i++) {
             const renderPass = renderPasses[i];
+            const renderTarget = renderPass.renderTarget;
+
+            // if using a target, or null which represents the default framebuffer
+            if (renderTarget !== undefined) {
```
Comment (slimbuck) [inline src/scene/frame-graph.js:45]:
```suggestion
            if (renderTarget !== null) {
```

**Review by slimbuck — COMMENTED** [review]
To infinity and beyond!

File: src/graphics/render-pass.js (hunk 2 above)
Comment (mvaligursky) [reply src/graphics/render-pass.js:154]: this would change behavior, this is not desirable

Comment (slimbuck) [reply src/graphics/render-pass.js:154]: the line above means it'll never be undefined

Comment (slimbuck) [reply src/graphics/render-pass.js:154]: did you mean `renderTarget !== undefined` perhaps?

File: src/graphics/graphics-device.js (hunk 1 above)
Comment (mvaligursky) [reply src/graphics/graphics-device.js:78]: this is an existing member, but it was never declared and just assigned to in few places, so I added it as a member with type. It's not directly used by render passes. I'll add comment stating it's currently assigned render target.

File: src/graphics/render-pass.js (hunk 2 above)
Comment (mvaligursky) [reply src/graphics/render-pass.js:154]: true!

File: src/graphics/render-pass.js [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+     */
+    clear = false;
+
+    /**
+     * True if the attachment needs to be stored after the render pass. False
+     * if it can be discarded.
+     * Note: This relates to the surface that is getting rendered to, and can be either
+     * single or multi-sampled. Further, it multi-sampled surface is used, the resolve
+     * flag further specifies if this gets resolved to a single-sampled surface. This
+     * behaviour matches WebGPU specification.
```
Comment (willeastcott) [inline src/graphics/render-pass.js:25]:
```suggestion
* behavior matches WebGPU specification.
```

File: src/graphics/render-pass.js [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+     * True if the attachment should be cleared before rendering, false to preserve
+     * the existing content.
+     */
+    clear = false;
+
+    /**
+     * True if the attachment needs to be stored after the render pass. False
+     * if it can be discarded.
+     * Note: This relates to the surface that is getting rendered to, and can be either
+     * single or multi-sampled. Further, it multi-sampled surface is used, the resolve
```
Comment (willeastcott) [inline src/graphics/render-pass.js:23]:
```suggestion
     * single or multi-sampled. Further, if a multi-sampled surface is used, the resolve

File: src/graphics/render-pass.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+class DepthStencilAttachmentOps {
+    /**
+     * A depth value used to clear the depth attachment when the clear is enabled.
+     */
+    clearDepthValue = 1;
+
+    /**
+     * A scencil value used to clear the stencil attachment when the clear is enabled.
```
Comment (willeastcott) [inline src/graphics/render-pass.js:53]:
```suggestion
     * A stencil value used to clear the stencil attachment when the clear is enabled.
```

File: src/graphics/render-pass.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    clearStencilValue = 0;
+
+    /**
+     * True if the depth attachment should be cleared before rendering, false to preserve
+     * the existing content.
+     */
+    clearDepth = false;
+
+    /**
+     * True if the stencil  attachment should be cleared before rendering, false to preserve
```
Comment (willeastcott) [inline src/graphics/render-pass.js:64]:
```suggestion
     * True if the stencil attachment should be cleared before rendering, false to preserve
```

File: src/graphics/render-pass.js [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+    /** @type {ColorAttachmentOps} */
+    colorOps;
+
+    /** @type {DepthStencilAttachmentOps} */
+    depthStencilOps;
+
+    /**
+     * If true, this pass might use dynamically rendered cubemaps. Use for a case where rendering to cubemap
+     * faces is interleaved with rendering to shadows, to avoid generating cubemap mipmaps. This will likely
+     * be retired modified when render target dependency tracking gets implemented.
```
Comment (willeastcott) [inline src/graphics/render-pass.js:116]: Retired or modified?

File: src/graphics/render-pass.js [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+     *
+     * @type {boolean}
+     */
+    requiresCubemaps = true;
+
+    /**
+     * True if the render pass uses full viewport / scissor for the rendering into the render target.
```
Comment (willeastcott) [inline src/graphics/render-pass.js:123]:
```suggestion
     * True if the render pass uses the full viewport / scissor for rendering into the render target.
```

**Review by willeastcott — APPROVED** [review]
Approving based on a layout + spelling + grammar pass. 😄  I'll leave a technical approval to @slimbuck.

File: src/graphics/render-pass.js (hunk 12 above)
Comment (mvaligursky) [reply src/graphics/render-pass.js:116]: I'll go with retired :)

File: src/graphics/render-pass.js (hunk 3 above)
Comment (mvaligursky) [reply src/graphics/render-pass.js:161]: added comment: if rendering to single-sampled buffer, this buffer needs to be stored

File: src/graphics/render-pass.js (hunk 4 above)
Comment (slimbuck) [reply src/graphics/render-pass.js:208]: Still wondering ab this one?

**Pull request merged**